### ISSUE TITLE
Identical file name path in tab is too long and difficult to diff

### ISSUE THREAD

**Issue by Tyriar** [Microsoft/vscode#12040]
- VSCode Version: Code - Insiders 1.6.0-insider (fad6a894c197ee791403ac694c9316b0654f31a4, 2016-09-14T13:18:39.616Z)
- OS Version: Linux x64 4.4.0-36-generic, Ubuntu 16.04

Current:

![image](https://cloud.githubusercontent.com/assets/2193314/18524039/a403a33c-7a6b-11e6-94e6-02f60bc61668.png)

Proposed:

![image](https://cloud.githubusercontent.com/assets/2193314/18524117/fe9a1f7e-7a6b-11e6-8c73-0d6aac733096.png)

To describe the behavior I want: It should replace all parts of the start of the path that are the same with `…/`, then add the differing word(s) and then either `/…/` or `/`. This makes it easier to differentiate each tab and also reduces the total space taken by the tab particularly in a deep directory structure.

Some interesting cases could arise when more than 2 files collide in which maybe 2 parts of the path would need to be displayed, for example:
- `src/vs/workbench/parts/execution/electron-browser/terminalService.ts`
- `src/vs/workbench/parts/execution/electron-browser/something/terminalService.ts`
- `src/vs/workbench/parts/terminal/electron-browser/terminalService.ts`

Could yield:
- `…/execution/electron-browser/terminalService.ts`
- `…/execution/electron-browser/something/terminalService.ts`
- `…/terminal/…/terminalService.ts`

Basically just providing enough information to uniquely identify the file, `electron-browser` could not be truncated above as `…/execution/…/terminalService.ts` represents both files.

Another interesting edge case: single letter folders would probably consume more space when truncated.

**Comment by bpasero** [comment]
Sounds like a fun PR. The entry point is https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/browser/parts/editor/tabsTitleControl.ts#L258

**Comment by kumarharsh** [comment]
Great idea, its confusing when 60% of my files are named `index.js` -_- although when with VSCode, this is less of an issue when not using tabs.
Just a minor point: I rather like the file name on the left-most side of the tab. Perhaps overflowing the folder name on the left side instead of the right would be more helpful?

**Comment by Tyriar** [comment]
@kumarharsh do you mean like in my proposal or like `terminalService.ts - …/terminal/…/`?

**Comment by luisa123456** [comment]
Long Path Tool is the perfect utility that can solve all the long path issues. Download this software online.

**Comment by kumarharsh** [comment]
The second part of my comment:

>  I rather like the file name on the left-most side of the tab. Perhaps overflowing the folder name on the left side instead of the right would be more helpful?

was meant to provide an example of what is **good** in the current UI (filename first, folder low-lighted). Perhaps if the UI can somehow merge the goodness of both the existing UI as well as your proposal (unique folder names, lowlighted)...

**Comment by mac-pro** [comment]
I would suggest this one for your example:

-    …/electron-browser/terminalService.ts
-    …/something/terminalService.ts
-    …/terminal/…/terminalService.ts

I mean: Showing the left-first different names is good enough. When the user points the mouse at the tabs, VSC already shows up full paths.

Another idea: you can add a second line and put the path there. That will leave more room. It could be an option in settings (whether to use second line or not).

**Comment by Tyriar** [comment]
@mac-pro but the last one is not enough information imo as both `…/electron-browser/terminalService.ts` and `…/terminal/…/terminalService.ts` could mean `src/vs/workbench/parts/terminal/electron-browser/terminalService.ts`.

**Comment by hun1ahpu** [comment]
As I understand ... represents one or more path segments?
So a/b/c/d.ts and f/b/c/d.ts will be replaced by a/.../d.ts and f/.../d.ts ?

**Comment by Tyriar** [comment]
@hun1ahpu yes, provided a and f dirs don't appear in each others paths elsewhere. Enough information to disambiguate the file name.

**Comment by bpasero** [comment]
* [ ] **To verify:** 
Verify that you see meaningful descriptions (those things besides the main label) when opening files with the same name but inside different folders. Our VS Code sources are a good candidate for testing (e.g. terminalxy).

**Comment by Tyriar** [comment]
Verified with 2 minor issues filed, thanks again @hun1ahpu :+1:
